group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorService.java

```diff
@@ -428,9 +430,44 @@ public CompletableFuture<ListGroupsResponseData> listGroups(
             return FutureUtils.failedFuture(Errors.COORDINATOR_NOT_AVAILABLE.exception());
         }
 
-        return FutureUtils.failedFuture(Errors.UNSUPPORTED_VERSION.exception(
-            "This API is not implemented yet."
-        ));
+        List<CompletableFuture<ListGroupsResponseData>> futures = new java.util.ArrayList<>(Collections.emptyList());
+        for (TopicPartition tp : runtime.partitions()) {
+            futures.add(runtime.scheduleReadOperation(
+                "list_groups",
+                tp,
+                (coordinator, lastCommittedOffset) -> coordinator.listGroups(context, request, lastCommittedOffset)
+            ).exceptionally(exception -> {
+                if (!(exception instanceof KafkaException)) {
+                    log.error("ListGroups request {} hit an unexpected exception: {}",
+                        request, exception.getMessage());
+                }
+                return new ListGroupsResponseData()
+                    .setErrorCode(Errors.forException(exception).code());
+            }));
```

Comment:
I think that we need to think a little more about the different errors that we could get here. My understanding is that we fail the entire requests in two cases: 1) at least one partition is loading; and 2) there is an unexpected error (non KafkaException).

Here, we could get the following errors:
- NotCoordinatorException if `tp` is no longer active, failed, etc. In this case, we actually want to return an empty lit of groups.
- CoordinatorLoadingException if `tp` is being loaded. In this case, we want to fail the entire request.
- Unexpected Exception. In this case, we also want to fail the entire request.

Knowing this, we should explicitly handle the NotCoordinatorException case here. For the other cases, would it be possible to re-throw the exception?

It would be great if you could also add a few unit tests for this.

Reply:
done.